test/unit/spans/batch-span-streamer.test.js

```diff
@@ -0,0 +1,232 @@
+    spanStreamer.queue_size = 1
+    const metrics = spanStreamer._metrics
+
+    t.equal(spanStreamer.spans.length, 0, 'no spans queued')
+
+    const fakeSpan = new SpanStreamerEvent('sandwich', {}, {})
+    const fakeSpanQueued = new SpanStreamerEvent('porridge', {}, {})
+
+    spanStreamer.write(fakeSpan)
+    spanStreamer.write(fakeSpanQueued)
+
+    t.equal(spanStreamer.spans.length, 1, 'one span queued')
+
+    // emit drain event, allow writes and check for span.trace_id
+    fakeConnection.stream.emit(
+      'drain',
+      (fakeConnection.stream.write = ({ spans }) => {
+        const [span] = spans
+        t.equal(span.trace_id, 'porridge', 'Should have formatted span')
+
+        return true
+      })
+    )
+
+    t.equal(spanStreamer.spans.length, 0, 'drained spans')
+    t.equal(
+      metrics.getOrCreateMetric(METRIC_NAMES.INFINITE_TRACING.DRAIN_DURATION).callCount,
+      1,
+      'DRAIN_DURATION metric'
+    )
+
+    t.equal(
+      metrics.getOrCreateMetric(METRIC_NAMES.INFINITE_TRACING.SENT).callCount,
+      2,
+      'SENT metric incremented'
+    )
+
+    t.end()
+  })
+
+  t.test('should send a batch if it exceeds queue', (t) => {
+    t.plan(11)
+    const metrics = spanStreamer._metrics
+
+    let i = 0
+    fakeConnection.stream.write = ({ spans }) => {
+      i++
+      if (i === 1) {
+        const [span, span2] = spans
+        t.equal(span.trace_id, 'sandwich', 'batch 1 span 1 ok')
+        t.equal(span2.trace_id, 'porridge', 'batch 1 span 2 ok')
+      } else {
+        const [span, span2] = spans
+        t.equal(span.trace_id, 'arepa', 'batch 2 span 1 ok')
+        t.equal(span2.trace_id, 'hummus', 'batch 2 span 2 ok')
+      }
+
+      return true
+    }
+
+    t.equal(spanStreamer.spans.length, 0, 'no spans queued')
+
+    const fakeSpan = new SpanStreamerEvent('sandwich', {}, {})
+    const fakeSpan2 = new SpanStreamerEvent('porridge', {}, {})
+    const fakeSpan3 = new SpanStreamerEvent('arepa', {}, {})
+    const fakeSpan4 = new SpanStreamerEvent('hummus', {}, {})
+
+    spanStreamer.write(fakeSpan)
+    t.equal(spanStreamer.spans.length, 1, '1 span in queue')
+
+    spanStreamer.write(fakeSpan2)
+
+    t.equal(spanStreamer.spans.length, 0, '0 spans in queue')
+    t.equal(
+      metrics.getOrCreateMetric(METRIC_NAMES.INFINITE_TRACING.SENT).callCount,
+      2,
+      'SENT metric incremented to 2'
+    )
+
+    spanStreamer.write(fakeSpan3)
+
+    t.equal(spanStreamer.spans.length, 1, '1 span in queue')
+
+    spanStreamer.write(fakeSpan4)
+
+    t.equal(spanStreamer.spans.length, 0, '0 spans in queue')
+    t.equal(
+      metrics.getOrCreateMetric(METRIC_NAMES.INFINITE_TRACING.SENT).callCount,
+      4,
+      'SENT metric incremented to 4'
+    )
+  })
+
+  // this will simulate n full batches and the last being 1/3 full
+  t.test('should send in appropriate batch sizes', (t) => {
+    const SPANS = 10000
+    const BATCH = 750
+    t.plan(Math.ceil(SPANS / BATCH) + 1)
```

Comment:
It's setting the number of assertions to the number of batches + the extract assertion on the sent metric. I'm setting the plan as I want to ensure all assertions are happening on stream as the writes are async and a little wonky to turn the stream into a a promise

Reply:
> It seems to be just as fine without it (as well as above).

Maybe but I don't want to make the assumption the stream will assert data before the test ends because it's async. this is just being more explicit

Reply:
Ah, I see that `plan` also measures asserts, not just subtests, thanks.